src/algorithms/forwardbackward.jl

```diff
@@ -0,0 +1,162 @@
+using Base.Iterators
+using ProximalAlgorithms.IterationTools
+using ProximalOperators: Zero
+using LinearAlgebra
+using Printf
+
+struct FBS_iterable{R <: Real, C <: Union{R, Complex{R}}, Tx <: AbstractArray{C}, Tf, TA, Tg}
+    f::Tf             # smooth term
+    A::TA             # matrix/linear operator
+    g::Tg             # (possibly) nonsmooth, proximable term
+    x0::Tx            # initial point
+    gamma::Maybe{R}   # stepsize parameter of forward and backward steps
+    adaptive::Bool    # enforce adaptive stepsize even if L is provided
+    fast::Bool
+end
+
+mutable struct FBS_state{R <: Real, Tx, TAx}
+    x::Tx             # iterate
+    Ax::TAx           # A times x
+    f_Ax::R           # value of smooth term
+    grad_f_Ax::Tx     # gradient of f at Ax
+    At_grad_f_Ax::TAx # gradient of smooth term
+    gamma::R          # stepsize parameter of forward and backward steps
+    y::Tx             # forward point
+    z::Tx             # forward-backward point
+    g_z::R            # value of nonsmooth term (at z)
+    res::Tx           # fixed-point residual at iterate (= z - x)
+    theta::R
+    z_prev::Tx
+end
+
+f_model(state::FBS_state) = f_model(state.f_Ax, state.At_grad_f_Ax, state.res, state.gamma)
+
+function Base.iterate(iter::FBS_iterable{R}) where R
+    x = iter.x0
+    Ax = iter.A * x
+    grad_f_Ax, f_Ax = gradient(iter.f, Ax)
+
+    gamma = iter.gamma
+
+    if gamma === nothing
+        # compute lower bound to Lipschitz constant of the gradient of x ↦ f(Ax)
+        xeps = x .+ R(1)
+        grad_f_Axeps, f_Axeps = gradient(iter.f, iter.A*xeps)
+        L = norm(iter.A' * (grad_f_Axeps - grad_f_Ax)) / R(sqrt(length(x)))
+        gamma = R(1)/L
+    end
+
+    # compute initial forward-backward step
+    At_grad_f_Ax = iter.A' * grad_f_Ax
+    y = x - gamma .* At_grad_f_Ax
+    z, g_z = prox(iter.g, y, gamma)
+
+    # compute initial fixed-point residual
+    res = x - z
+
+    state = FBS_state(x, Ax, f_Ax, grad_f_Ax, At_grad_f_Ax, gamma, y, z, g_z, res, R(1), copy(x))
+
+    return state, state
+end
+
+function Base.iterate(iter::FBS_iterable{R}, state::FBS_state{R, Tx, TAx}) where {R, Tx, TAx}
+    Az, f_Az, grad_f_Az, At_grad_f_Az = nothing, nothing, nothing, nothing
```

Comment:
but this could be initialised already when constructing `state` in `Base.iterate(iter::FBS_iterable{R})`? 
Or am I wrong?

Reply:
You mean having additional fields for these in the `state`? Yes, that could be, that could save some allocations I guess?

Reply:
I think I've addressed everything except for this: I would open a separate issue and leave it to a separate PR, from what I've seen algorithms perform already very good like this